substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1062 @@
+			slot_idx -= epoch_len;
+			if tickets_meta.unsorted_tickets_count != 0 {
+				Self::sort_tickets(u32::MAX, epoch_tag, &mut tickets_meta);
+				TicketsMeta::<T>::set(tickets_meta);
+			}
+		} else if slot_idx >= 2 * epoch_len {
+			return None
+		}
+
+		let ticket_idx = get_ticket_idx(slot_idx);
+		if ticket_idx < tickets_meta.tickets_count[epoch_tag as usize] {
+			TicketsIds::<T>::get((epoch_tag, ticket_idx))
+		} else {
+			None
+		}
+	}
+
+	/// Returns ticket id and data associated to the given `slot`.
+	///
+	/// Refer to the `slot_ticket_id` documentation for the slot-ticket association
+	/// criteria.
+	pub fn slot_ticket(slot: Slot) -> Option<(TicketId, TicketBody)> {
+		Self::slot_ticket_id(slot).and_then(|id| TicketsData::<T>::get(id).map(|body| (id, body)))
+	}
+
+	// Lexicographically sort the tickets which belong to the next epoch.
+	//
+	// Tickets are fetched from at most `max_segments` segments.
+	//
+	// The resulting sorted vector is optionally truncated to contain at most `MaxTickets`
+	// entries. If all the unsorted segments are consumed then the sorted vector is
+	// saved as the next epoch tickets, else it is saved to be used by next calls to
+	// this function.
+	pub(crate) fn sort_tickets(max_segments: u32, epoch_tag: u8, metadata: &mut TicketsMetadata) {
+		let mut unsorted_segments_count =
+			metadata.unsorted_tickets_count.div_ceil(SEGMENT_MAX_SIZE);
+		let max_segments = max_segments.min(unsorted_segments_count);
+		let max_tickets = MaxTicketsFor::<T>::get() as usize;
+
+		// Fetch the sorted candidates (if any).
+		let mut sorted_candidates = SortedCandidates::<T>::take().into_inner();
+
+		let mut require_sort = max_segments != 0;
+
+		// There is an upper bound to check only if we already sorted the max number
+		// of allowed tickets.
+		let mut upper_bound = *sorted_candidates.get(max_tickets - 1).unwrap_or(&TicketId::MAX);
+
+		// Consume at most `max_segments` segments.
+		// During the process remove every stale ticket from `TicketsData` storage.
+		for _ in 0..max_segments {
+			unsorted_segments_count -= 1;
+			let segment = UnsortedSegments::<T>::take(unsorted_segments_count);
+			metadata.unsorted_tickets_count -= segment.len() as u32;
+
+			// Push only ids with a value less than the current `upper_bound`.
+			// As ticket ids follow a uniform random distribution we expect to
+			// drop more or less half of the segment tickets here.
+			segment.into_iter().for_each(|id| {
+				if id < upper_bound {
+					sorted_candidates.push(id);
+				} else {
+					TicketsData::<T>::remove(id);
+				}
+			});
+
+			if sorted_candidates.len() > max_tickets {
+				// Sort, truncate good tickets, cleanup storage.
+				require_sort = false;
+				sorted_candidates.sort_unstable();
+				sorted_candidates[max_tickets..]
+					.iter()
+					.for_each(|id| TicketsData::<T>::remove(id));
+				sorted_candidates.truncate(max_tickets);
+				upper_bound = sorted_candidates[max_tickets - 1];
+			}
+		}
+
+		if require_sort {
```

Comment:
nit: how about adding segment_id() (and other related to indexing logic) method to Metadata?